tl

```diff
@@ -391,6 +411,109 @@ for _, include in ipairs(tlconfig["include_dir"]) do
+      :gsub(path_separator .. "+", path_separator))
+      :gsub(path_separator .. "+$", "")
+end
+
+local function path_concat(...)
+   local path = {}
+   for i = 1, select("#", ...) do
+      local fname = cleanup_file_name((select(i, ...)))
+      if #fname > 0 then
+         table.insert(path, fname)
+      end
+   end
+   return table.concat(path, path_separator)
+end
+
+local function remove_leading_path(leading_part, path)
+   local s, e = path:find("^" .. leading_part .. path_separator .. "?")
+   if s then
+      return path:sub(e+1, -1)
+   end
+   return path
+end
+
+local function traverse(dirname, emptyref,is_generated,generated_ref)
+   local files = {}
+   local paths = {} --lookup table for string paths to help
+   -- with pattern matching while iterating over a project
+   -- paths[files.foo.bar] -> "foo/bar"
+   emptyref = emptyref or {}
+   generated_ref = generated_ref or {}
+   for file in lfs.dir(dirname) do
+      if file ~= "." and file ~= ".." then
+         if lfs.attributes(path_concat(dirname, file), "mode") == "directory" then
+            local p
+            local prefix = dirname
+            files[file], p = traverse(path_concat(dirname, file), emptyref,is_generated,generated_ref)
+            paths[files[file]] = file
+            for k, v in pairs(p) do
+               paths[k] = path_concat(file, v)
+            end
+         else
+            -- storing a special entry in this table to it mark as empty could
+            -- interfere with convoluted or maliciously constructed directory
+            -- names so we use a table with specific metatable to mark
+            -- something as the end of a traversal to have a property attached
+            -- to the table, without creating an entry in the table
+            local meta_table = {empty=emptyref}
+            if is_generated then
+               meta_table["generated"] = generated_ref
+            end
+            files[file] = setmetatable({}, meta_table)
+            paths[files[file]] = file
+         end
+      end
+   end
+   return files, paths, emptyref
+end
+
+local function delete_dir(dir)
+   for file in lfs.dir(dir) do
+      if file ~= "." and file ~= ".." then 
+         local file_path = path_concat(dir,file)
+         if file ~= "." and file ~= ".." then
+            if lfs.attributes(file_path, 'mode') == 'file' then
+                  os.remove(file_path)
+            elseif lfs.attributes(file_path, 'mode') == 'directory' then
+                  delete_dir(file_path)
+            end
+         end
+      end
+   end
+   lfs.rmdir(dir)
+end
+
+local build_path
+
+local function cleanup_generated_files()
+   if build_path then
+      delete_dir(build_path)
+   end
+end
+if tlconfig["build_file"] then
+   local tmp_dir = os.getenv("TMP") or os.getenv("TMPDIR") or "/tmp" --fall back to /tmp, apparently some (mine?) computers don't have a variable set for it.
+   build_path = path_concat(tmp_dir,"tl_temp_out" .. os.time(os.date("!*t")))
```

Comment:
> os.time(os.date("!*t"))

super minor thing because I happened to look at this bit of the code because of the comment above: I think `os.time()` with no arguments does the same?

Reply:
it could be, I searched how to get a unix timestamp (as I forgot) and this answer showed up first but now that you mention it, I am pretty sure as well that just `os.time()` is good enough as well.

Reply:
just checked it

> The time function, when called without arguments, returns the current date and time, coded as a number. (In most systems, that number is the number of seconds since some epoch.) 

https://www.lua.org/pil/22.1.html

I think the keyword is "most" in this case. However, as it is only used to reduce the chance of collisions (Especially with a previous run that couldn't delete the folder for one reason or another) I think that using `os.time` without arguments is still fine, for as long as those other ways have at least second precision.

Reply:
> https://www.lua.org/pil/22.1.html

I think the above link is correct, but beware in general that lua.org/pil refers to Lua 5.0, which is out of date; usually only 5.1+ matter.

Given the [latest docs](https://www.lua.org/manual/5.4/manual.html#pdf-os.time) (which haven't changed much since 5.1), I think the outputs of `os.time` are just as system-dependent with or without arguments.